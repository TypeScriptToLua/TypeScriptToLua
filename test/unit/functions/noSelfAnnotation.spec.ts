import * as util from "../../util";

const methodHolders = ["class", "interface"];

test("@noSelf on declared function removes context argument", () => {
    util.testModule`
        /** @noSelf */
        declare function myFunction(): void;
        myFunction();
    `.expectLuaToMatchSnapshot();
});

test.each(methodHolders)("@noSelf on method inside %s declaration removes context argument", holderType => {
    util.testModule`
        declare ${holderType} MethodHolder {
            /** @noSelf */
            myMethod(): void;
        }
        declare const holder: MethodHolder;
        holder.myMethod();
    `.expectLuaToMatchSnapshot();
});

test.each(methodHolders)("@noSelf on parent %s declaration removes context argument", holderType => {
    util.testModule`
        /** @noSelf */
        declare ${holderType} MethodHolder {
            myMethod(): void;
        }
        declare const holder: MethodHolder;
        holder.myMethod();
    `.expectLuaToMatchSnapshot();
});

test("@noSelf on method inside namespace declaration removes context argument", () => {
    util.testModule`
        declare namespace MyNamespace {
            /** @noSelf */
            export function myMethod(): void;
        }
        MyNamespace.myMethod();
    `.expectLuaToMatchSnapshot();
});

test("@noSelf on parent namespace declaration removes context argument", () => {
    util.testModule`
        /** @noSelf */
        declare namespace MyNamespace {
            export function myMethod(): void;
        }
        MyNamespace.myMethod();
    `.expectLuaToMatchSnapshot();
});

test("@noSelf on static class methods with string key access", () => {
    util.testModule`
        /** @noSelf */
        declare class TestClass {
            static [key: string]: () => void;
            static myMethod(): void;
        }

        TestClass.myMethod();
        TestClass.myKey();
    `.expectLuaToMatchSnapshot();
});

// additional coverage for https://github.com/TypeScriptToLua/TypeScriptToLua/issues/1292
test("explicit this parameter respected over @noSelf", () => {
    util.testModule`
        /** @noSelfInFile */
        function foo(this: unknown, arg: any) {
            return {self: this, arg};
        }
        export const result = foo(1);
    `.expectToMatchJsResult();
});

test("respect noSelfInFile over noImplicitSelf", () => {
    const result = util.testModule`
        /** @noSelfInFile **/
        const func: Function = () => 1;
        export const result = func(1);
    `
        .expectToMatchJsResult()
        .getLuaResult();

    expect(result.transpiledFiles).not.toHaveLength(0);

    const mainFile = result.transpiledFiles.find(f => f.outPath === "main.lua");
    expect(mainFile).toBeDefined();

    // avoid ts error "not defined", even though toBeDefined is being checked above
    if (!mainFile) return;

    expect(mainFile.lua).toBeDefined();
    expect(mainFile.lua).toContain("func(1)");
});

test("respect noSelfInFile over noImplicitSelf (func declared in other file)", () => {
    const result = util.testModule`
        import { func, result } from "./functions";

        export const result1 = result;
        export const result2 = func(1);
    `
        .addExtraFile(
            "functions.ts",
            `
            /** @noSelfInFile **/
            export const func: Function = () => 1;
            export const result = func(2);
            `
        )
        .expectToMatchJsResult()
        .getLuaResult();

    expect(result.transpiledFiles).not.toHaveLength(0);

    const mainFile = result.transpiledFiles.find(f => f.outPath.includes("main.lua"));
    expect(mainFile).toBeDefined();
    const functionFile = result.transpiledFiles.find(f => f.outPath.includes("functions.lua"));
    expect(functionFile).toBeDefined();

    // avoid ts error "not defined", even though toBeDefined is being checked above
    if (!mainFile || !functionFile) return;

    expect(mainFile.lua).toBeDefined();
    expect(mainFile.lua).toContain("func(1)");
    expect(mainFile.lua).toBeDefined();
    expect(functionFile.lua).toContain("func(2)");
});
