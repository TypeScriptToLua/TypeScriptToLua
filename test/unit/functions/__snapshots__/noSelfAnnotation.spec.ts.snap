// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@noSelf on declared function removes context argument 1`] = `"myFunction()"`;

exports[`@noSelf on method inside class declaration removes context argument 1`] = `"holder.myMethod()"`;

exports[`@noSelf on method inside interface declaration removes context argument 1`] = `"holder.myMethod()"`;

exports[`@noSelf on method inside namespace declaration removes context argument 1`] = `"MyNamespace.myMethod()"`;

exports[`@noSelf on parent class declaration removes context argument 1`] = `"holder.myMethod()"`;

exports[`@noSelf on parent interface declaration removes context argument 1`] = `"holder.myMethod()"`;

exports[`@noSelf on parent namespace declaration removes context argument 1`] = `"MyNamespace.myMethod()"`;

exports[`@noSelf on static class methods with string key access 1`] = `
"TestClass.myMethod()
TestClass.myKey()"
`;
